packages/runtime/src/runtime/hooks/useEnsembleAction.tsx

```diff
@@ -299,6 +315,179 @@ export const useInvokeAPI: EnsembleActionHook<InvokeAPIAction> = (action) => {
   return invokeApi;
 };
 
+export const useConnectSocket: EnsembleActionHook<ConnectSocketAction> = (
+  action,
+) => {
+  const { sockets, setData } = useScreenData();
+  const [isComplete, setIsComplete] = useState<boolean>();
+  const [isLoading, setIsLoading] = useState<boolean>();
+
+  const socket = useMemo(
+    () => sockets?.find((model) => model.name === action?.name),
+    [action, sockets],
+  );
+
+  const onSocketConnectAction = useEnsembleAction(socket?.onSuccess);
+  const onMessageReceiveAction = useEnsembleAction(socket?.onReceive);
+  const onSocketDisconnectAction = useEnsembleAction(socket?.onDisconnect);
+
+  const connectSocket = useMemo(() => {
+    if (!socket) {
+      return;
+    }
+
+    const callback = (): void => {
+      setIsComplete(false);
+    };
+    return { callback };
+  }, [socket]);
+
+  useEffect(() => {
+    if (!socket || isComplete !== false || isLoading) {
+      return;
+    }
+
+    const fireRequest = (): void => {
+      setIsLoading(true);
+      try {
+        const ws = new WebSocket(socket.uri);
+
+        ws.onopen = (): void => {
+          onSocketConnectAction?.callback();
+        };
+
+        ws.onmessage = (e: MessageEvent): void => {
+          onMessageReceiveAction?.callback({ data: e.data as unknown });
+        };
+
+        ws.onclose = (): void => {
+          onSocketDisconnectAction?.callback();
+        };
+
+        setData(socket.name, ws);
+      } catch (e) {
+        logError(e);
+      } finally {
+        setIsLoading(false);
+        setIsComplete(true);
```

Comment:
These should be set in the event hooks on the socket itself. All this is doing is created a Websocket object but it's not doing the connection.